urnai/sc2/states/terran_state.py

```diff
@@ -0,0 +1,121 @@
+
+
+class TerranState(StateBase):
+    
+    def __init__(self, grid_size=4):
+
+        self.grid_size = grid_size
+        # Size of the state returned
+        # 22: number of data added to the state (number of minerals, army_count, etc)
+        # 2 * 4 ** 2: 2 grids of size 4 x 4, representing enemy and player units
+        self._state_size = int(22 + 2 * (self.grid_size ** 2))
+        self.player_race = sc2_env.Race.terran
+        self.base_top_left = None
+        self._state = None
+
+    def update(self, obs):
+        if obs.game_loop[0] < 80 and self.base_top_left is None:
+
+            commandcenter = get_units_by_type(obs, units.Terran.CommandCenter)
+
+            if len(commandcenter) > 0:
+                townhall = commandcenter[0]
+                self.player_race = sc2_env.Race.terran
+
+            self.base_top_left = (townhall.x < 32)
+
+        new_state = []
+        # Adds general information from the player.
+        new_state.append(obs.player.minerals / 6000)
+        new_state.append(obs.player.vespene / 6000)
+        new_state.append(obs.player.food_cap / 200)
+        new_state.append(obs.player.food_used / 200)
+        new_state.append(obs.player.food_army / 200)
+        new_state.append(obs.player.food_workers / 200)
+        new_state.append((obs.player.food_cap - obs.player.food_used) / 200)
+        new_state.append(obs.player.army_count / 200)
+        new_state.append(obs.player.idle_worker_count / 200)
+
+        # Adds information related to player's Terran units/buildings.
+        new_state.append(get_my_units_amount(obs, units.Terran.CommandCenter) +
+                        get_my_units_amount(obs, units.Terran.OrbitalCommand) +
+                        get_my_units_amount(obs, units.Terran.PlanetaryFortress) / 2)
+        new_state.append(get_my_units_amount(obs, units.Terran.SupplyDepot) / 18)
+        new_state.append(get_my_units_amount(obs, units.Terran.Refinery) / 4)
+        new_state.append(get_my_units_amount(obs, units.Terran.EngineeringBay))
+        new_state.append(get_my_units_amount(obs, units.Terran.Armory))
+        new_state.append(get_my_units_amount(obs, units.Terran.MissileTurret) / 4)
+        new_state.append(get_my_units_amount(obs, units.Terran.SensorTower)/1)
+        new_state.append(get_my_units_amount(obs, units.Terran.Bunker)/4)
+        new_state.append(get_my_units_amount(obs, units.Terran.FusionCore))
+        new_state.append(get_my_units_amount(obs, units.Terran.GhostAcademy))
+        new_state.append(get_my_units_amount(obs, units.Terran.Barracks) / 3)
+        new_state.append(get_my_units_amount(obs, units.Terran.Factory) / 2)
+        new_state.append(get_my_units_amount(obs, units.Terran.Starport) / 2)
+
+        # Instead of making a vector for all coordnates on the map, we'll
+        # discretize our enemy space
+        # and use a 4x4 grid to store enemy positions by marking a square as 1 if
+        # there's any enemy on it.
```

Comment:
The following piece of code could be in a separate function and only invoked here.

Reply:
About the following piece of code, I figured some possible mistakes. For example here:

```python
for i in range(0, len(enemy_units)):
            y = int(math.ceil((enemy_units[i].x + 1) / 64 / self.grid_size))
            x = int(math.ceil((enemy_units[i].y + 1) / 64 / self.grid_size))
            enemy_grid[x - 1][y - 1] += 1
```

I think the number `64` is for the current size of the map. So ideally, it shouldn't be a constant number, but an attribute from "obs" that represents the size of the map. 
Also, the parenthesis are incorrect, it should be like this:

```python
for i in range(0, len(enemy_units)):
            y = int(math.ceil((enemy_units[i].x + 1) / (64 / self.grid_size)))
            x = int(math.ceil((enemy_units[i].y + 1) / (64 / self.grid_size)))
            enemy_grid[x - 1][y - 1] += 1
```

Reply:
If 64 really does concern the size of the map, then we should remove it to make the class as generic as possible.

Reply:
In #83 , the observation returned from our Environment Class has an attribute called "map_size", which would give what is needed to replace this `64` number. However, it would be better to discover the map's size from what is already returned by the pysc2 pure observation, right?

Reply:
> However, it would be better to discover the map's size from what is already returned by the pysc2 pure observation, right?

Yes, and I believe it's the ideal way forward if possible.

Reply:
> Yes, and I believe it's the ideal way forward if possible.

Unfortunately, I don't think that's possible. Currently, this number corresponds to the `raw_resolution` of the map. In that case, only the `environment` class form pysc2 has access to it, and it doesn't add it as an attribute to a default `observation`. That said, the way I figured to solve this was to receive this `raw_resolution` as a parameter in the constructor of the `State` class.